File: YALMIP-master/mosek/11.0/tools/examples/dotnet/mioinitsol.cs
/*
   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.

   File:      mioinitsol.cs

   Purpose:   Demonstrates how to solve a MIP with a start guess.

 */
using System;

namespace mosek.example
{
  class msgclass : mosek.Stream
  {
    string prefix;
    public msgclass (string prfx)
    {
      prefix = prfx;
    }

    public override void streamCB (string msg)
    {
      Console.Write ("{0}{1}", prefix, msg);
    }
  }

  public class mioinitsol
  {
    public static void Main ()
    {
      // Since the value infinity is never used, we define
      // 'infinity' symbolic purposes only
      double infinity = 0;

      int numvar = 4;
      int numcon = 1;
      int NUMINTVAR = 3;

      double[] c = { 7.0, 10.0, 1.0, 5.0 };

      mosek.boundkey[] bkc = {mosek.boundkey.up};
      double[] blc = { -infinity};
      double[] buc = {2.5};
      mosek.boundkey[] bkx = {mosek.boundkey.lo,
                              mosek.boundkey.lo,
                              mosek.boundkey.lo,
                              mosek.boundkey.lo
                             };
      double[] blx = {0.0,
                      0.0,
                      0.0,
                      0.0
                     };
      double[] bux = {infinity,
                      infinity,
                      infinity,
                      infinity
                     };

      int[] ptrb = {0, 1, 2, 3};
      int[] ptre = {1, 2, 3, 4};
      double[] aval = {1.0, 1.0, 1.0, 1.0};
      int[] asub = {0, 0, 0, 0};
      int[] intsub = {0, 1, 2};

      try
      {
        // Create a task object
        using (var task = new mosek.Task()) 
        {        
          // Directs the log task stream to the user specified
          // method task_msg_obj.streamCB
          task.set_Stream (mosek.streamtype.log, new msgclass (""));
          task.inputdata(numcon, numvar,
                         c,
                         0.0,
                         ptrb,
                         ptre,
                         asub,
                         aval,
                         bkc,
                         blc,
                         buc,
                         bkx,
                         blx,
                         bux);

          for (int j = 0 ; j < NUMINTVAR ; ++j)
            task.putvartype(intsub[j], mosek.variabletype.type_int);
          task.putobjsense(mosek.objsense.maximize);

          // Assign values to integer variables.
          // We only set a slice of xx
          double[] values = {1.0, 1.0, 0.0};
          task.putxxslice(mosek.soltype.itg, 0, 3, values);

          // Request constructing the solution from integer variable values
          task.putintparam(mosek.iparam.mio_construct_sol, mosek.onoffkey.on);

          try
          {
            task.optimize();
            task.solutionsummary(mosek.streamtype.log);
          }
          catch (mosek.Warning w)
          {
            Console.WriteLine("Mosek warning:");
            Console.WriteLine (w.Code);
            Console.WriteLine (w);
          }
          
          double[] xx = task.getxx(mosek.soltype.itg);

          Console.WriteLine("Solution:");
          for (int j = 0; j < numvar; ++j)
            Console.WriteLine ("x[{0}]:{1}", j, xx[j]);

          // Was the initial solution used?
          int constr = task.getintinf(mosek.iinfitem.mio_construct_solution);
          double constrVal = task.getdouinf(mosek.dinfitem.mio_construct_solution_obj);
          Console.WriteLine("Construct solution utilization: " + constr);
          Console.WriteLine("Construct solution objective: " +  constrVal);
        }
      }
      catch (mosek.Exception e)
      {
        Console.WriteLine (e.Code);
        Console.WriteLine (e);
        throw;
      }
    }
  }
}

File: YALMIP-master/mosek/11.0/tools/examples/python/mioinitsol.py
##
#  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#  File :      mioinitsol.py
#
#  Purpose :   Demonstrates how to solve a small mixed
#              integer linear optimization problem using the MOSEK Python API.
##
import sys
import mosek

# Since the actual value of Infinity is ignores, we define it solely
# for symbolic purposes:
inf = 0.0

# Define a stream printer to grab output from MOSEK
def streamprinter(text):
    sys.stdout.write(text)
    sys.stdout.flush()


def main():
    # Create a task
    with mosek.Task() as task:
        # Attach a printer to the task
        task.set_Stream(mosek.streamtype.log, streamprinter)

        bkc = [mosek.boundkey.up]
        blc = [-inf, ]
        buc = [2.5]

        bkx = [mosek.boundkey.lo,
               mosek.boundkey.lo,
               mosek.boundkey.lo,
               mosek.boundkey.lo]

        blx = [0.0, 0.0, 0.0, 0.0]
        bux = [inf, inf, inf, inf]

        c = [7.0, 10.0, 1.0, 5.0]

        asub = [0, 0, 0, 0]
        acof = [1.0, 1.0, 1.0, 1.0]

        ptrb = [0, 1, 2, 3]
        ptre = [1, 2, 3, 4]

        numvar = len(bkx)
        numcon = len(bkc)

        # Input linear data
        task.inputdata(numcon, numvar,
                       c, 0.0,
                       ptrb, ptre, asub, acof,
                       bkc, blc, buc,
                       bkx, blx, bux)

        # Input objective sense
        task.putobjsense(mosek.objsense.maximize)

        # Define variables to be integers
        task.putvartypelist([0, 1, 2],
                            [mosek.variabletype.type_int]*3)

        # Assign values to integer variables. 
        # (We only set a slice of xx)
        task.putxxslice(mosek.soltype.itg, 0, 3, [1.0, 1.0, 0.0])

        # Request constructing the solution from integer variable values
        task.putintparam(mosek.iparam.mio_construct_sol, mosek.onoffkey.on)

        # Optimize
        task.optimize()
        task.writedata("mioinitsol.ptf")

        task.solutionsummary(mosek.streamtype.log)

        if task.solutiondef(mosek.soltype.itg):
            # Output a solution
            xx = task.getxx(mosek.soltype.itg)

            print("Integer optimal solution")
            for j in range(0, numvar):
                print("\tx[%d] = %e" % (j, xx[j]))

            # Was the initial guess used?
            constr = task.getintinf(mosek.iinfitem.mio_construct_solution)
            constrVal = task.getdouinf(mosek.dinfitem.mio_construct_solution_obj)
            print("Construct solution utilization: {0}\nConstruct solution objective: {1:.3f}\n".format(constr, constrVal))
        else:
            print("No integer solution is available.")


# call the main function
try:
    main()
except mosek.MosekException as e:
    print("ERROR: %s" % str(e.errno))
    if e.msg is not None:
        print("\t%s" % e.msg)
    sys.exit(1)
except:
    import traceback
    traceback.print_exc()
    sys.exit(1)
